In [7]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.preprocessing import LabelEncoder,StandardScaler
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.compose import ColumnTransformer, make_column_selector
from sklearn.impute import SimpleImputer

# Models
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.neural_network import MLPRegressor
import xgboost as xgb
import lightgbm as lgb

In [8]:
# train_data = pd.read_csv("a1_dataset/train_v9rqX0R.csv")
# test_data = pd.read_csv("a1_dataset/test_AbJTz2l.csv")

In [9]:
# X = train_data.drop("Item_Outlet_Sales", axis=1)
# y = train_data["Item_Outlet_Sales"]


# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=25)


# imputer = ColumnTransformer(
#     transformers=[
#         ("cat_imputer", SimpleImputer(strategy="most_frequent"), make_column_selector(dtype_include=object)),
#         ("num_imputer", SimpleImputer(strategy="mean"), make_column_selector(dtype_include=np.number))
#     ],
#     remainder="passthrough",
#     verbose_feature_names_out=False
# )

# encoder = ColumnTransformer(
#     transformers=[
#         ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False), make_column_selector(dtype_include=object))
#     ],
#     remainder="passthrough",
#     verbose_feature_names_out=False
# )

# scaler = StandardScaler()

# X_train_imputed = pd.DataFrame(imputer.fit_transform(X_train), columns=X_train.columns)
# X_test_imputed = pd.DataFrame(imputer.transform(X_test), columns=X_test.columns)

# X_train_encoded = pd.DataFrame(encoder.fit_transform(X_train_imputed))
# X_test_encoded = pd.DataFrame(encoder.transform(X_test_imputed))

# X_train_scaled = scaler.fit_transform(X_train_encoded)
# X_test_scaled = scaler.transform(X_test_encoded)


In [10]:
# models = {
#     "Linear Regression": LinearRegression(),
#     "Ridge Regression": Ridge(),
#     "Lasso Regression": Lasso(),
#     "ElasticNet Regression": ElasticNet(),
#     "Decision Tree": DecisionTreeRegressor(max_depth=5),
#     "Random Forest": RandomForestRegressor(max_depth = 5),
#     "Gradient Boosting": GradientBoostingRegressor(),
#     "Support Vector Regressor": SVR(max_iter=10),
#     "MLP Regressor": MLPRegressor(max_iter=500)
# }


# for name, model in models.items():
#     model.fit(X_train_scaled, y_train)
#     y_pred = model.predict(X_test_scaled)
#     mse = mean_squared_error(y_test, y_pred)
#     r2 = r2_score(y_test, y_pred)
#     print(f"{name}:")
#     print(f"  Mean Squared Error: {mse:.2f}")
#     print(f"  R² Score: {r2:.2f}")
#     print("-" * 40)

Support Vector Regressor:
  Mean Squared Error: 9081291.93
  R² Score: -2.02
----------------------------------------
MLP Regressor:
  Mean Squared Error: 6513250.07
  R² Score: -1.17
----------------------------------------

In [ ]:


# Load dataset
df = pd.read_csv('a1_dataset/train_v9rqX0R.csv')

# Drop identifier columns
df = df.drop(['Item_Identifier', 'Outlet_Identifier'], axis=1)

# Apply preprocessing
df = preprocess(df)

# Features and target
X = df.drop('Item_Outlet_Sales', axis=1)
y = df['Item_Outlet_Sales']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define models
models = {
    "Linear Regression": LinearRegression(),
    "Ridge Regression": Ridge(),
    "Lasso Regression": Lasso(),
    "ElasticNet Regression": ElasticNet(),
    "Decision Tree": DecisionTreeRegressor(max_depth=5),
    "Random Forest": RandomForestRegressor(n_estimators=100, random_state=42),
    "Gradient Boosting": GradientBoostingRegressor(n_estimators=100, random_state=42),
    "XGBoost": xgb.XGBRegressor(n_estimators=100, random_state=42, verbosity=0),
    "LightGBM": lgb.LGBMRegressor(n_estimators=100, random_state=42),
    "Support Vector Regressor": SVR(max_iter=100),
    "MLP Regressor": MLPRegressor(max_iter=500)
}

results = []

# Train and evaluate each model
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    r2 = r2_score(y_test, y_pred)
    results.append((name, rmse, r2))

# Sort results by RMSE
results.sort(key=lambda x: x[1])

# Print results
print("Model Performance (sorted by RMSE):")
for name, rmse, r2 in results:
    print(f"{name}: RMSE={rmse:.2f}, R2={r2:.4f}")

best_model = results[0][0]
print(f"\nBest model based on RMSE: {best_model}")

[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000068 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 777
[LightGBM] [Info] Number of data points in the train set: 6818, number of used features: 9
[LightGBM] [Info] Start training from score 2202.365232


C:\Users\nagar\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\svm\_base.py:305: ConvergenceWarning: Solver terminated early (max_iter=100).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


Model Performance (sorted by RMSE):
Decision Tree: RMSE=1027.79, R2=0.6113
Gradient Boosting: RMSE=1034.58, R2=0.6062
LightGBM: RMSE=1051.68, R2=0.5931
Random Forest: RMSE=1085.68, R2=0.5663
Lasso Regression: RMSE=1136.50, R2=0.5248
Ridge Regression: RMSE=1137.04, R2=0.5243
Linear Regression: RMSE=1137.22, R2=0.5242
MLP Regressor: RMSE=1144.40, R2=0.5182
XGBoost: RMSE=1145.41, R2=0.5173
ElasticNet Regression: RMSE=1201.22, R2=0.4691
Support Vector Regressor: RMSE=2185.06, R2=-0.7566

Best model based on RMSE: Decision Tree


C:\Users\nagar\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\sklearn\neural_network\_multilayer_perceptron.py:781: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


In [12]:


# Parameter grids
param_grids = {
    'DecisionTree': {
        'max_depth': [5, 10, 15],
        'min_samples_split': [2, 5, 10]
    },
    'GradientBoosting': {
        'n_estimators': [100, 200],
        'learning_rate': [0.05, 0.1],
        'max_depth': [3, 5]
    },
    'XGBoost': {
        'n_estimators': [100, 200],
        'learning_rate': [0.05, 0.1],
        'max_depth': [3, 5],
        'subsample': [0.8, 1.0]
    },
    'LightGBM': {
        'n_estimators': [100, 200],
        'learning_rate': [0.05, 0.1],
        'num_leaves': [31, 50]
    }
}

# Models
models = {
    'DecisionTree': DecisionTreeRegressor(random_state=42),
    'GradientBoosting': GradientBoostingRegressor(random_state=42),
    'XGBoost': xgb.XGBRegressor(random_state=42, verbosity=0),
    'LightGBM': lgb.LGBMRegressor(random_state=42)
}

results = []

# GridSearchCV for each model
for name, model in models.items():
    print(f"Tuning {name}...")
    grid = GridSearchCV(model, param_grids[name], cv=3, scoring='neg_root_mean_squared_error', n_jobs=-1)
    grid.fit(X_train, y_train)
    best_model = grid.best_estimator_
    y_pred = best_model.predict(X_test)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    r2 = r2_score(y_test, y_pred)
    results.append((name, rmse, r2, grid.best_params_))

# Sort by RMSE
results.sort(key=lambda x: x[1])

# Print results
print("\nTuned Model Performance (sorted by RMSE):")
for name, rmse, r2, params in results:
    print(f"{name}: RMSE={rmse:.2f}, R2={r2:.4f}, Best Params={params}")

best_model = results[0]
print(f"\nBest model after tuning: {best_model[0]} with RMSE={best_model[1]:.2f}, R2={best_model[2]:.4f}")
print(f"Best Parameters: {best_model[3]}")

Tuning DecisionTree...
Tuning GradientBoosting...
Tuning XGBoost...
Tuning LightGBM...
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000252 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 777
[LightGBM] [Info] Number of data points in the train set: 6818, number of used features: 9
[LightGBM] [Info] Start training from score 2202.365232

Tuned Model Performance (sorted by RMSE):
XGBoost: RMSE=1024.22, R2=0.6140, Best Params={'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 100, 'subsample': 1.0}
GradientBoosting: RMSE=1027.56, R2=0.6115, Best Params={'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 100}
DecisionTree: RMSE=1027.79, R2=0.6113, Best Params={'max_depth': 5, 'min_samples_split': 10}
LightGBM: RMSE=1035.57, R2=0.6054, Best Params={'learning_rate': 0.05, 'n_estimators': 100, 'num_leaves': 31}

Best model after tuning: XGBoost with RMSE=1024.22, R2=0.6140
Best Parameters: {'le

Tuned Model Performance (sorted by RMSE):
XGBoost: RMSE=1024.22, R2=0.6140, Best Params={'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 100, 'subsample': 1.0}
GradientBoosting: RMSE=1027.56, R2=0.6115, Best Params={'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 100}
DecisionTree: RMSE=1027.79, R2=0.6113, Best Params={'max_depth': 5, 'min_samples_split': 10}
LightGBM: RMSE=1035.57, R2=0.6054, Best Params={'learning_rate': 0.05, 'n_estimators': 100, 'num_leaves': 31}

Best model after tuning: XGBoost with RMSE=1024.22, R2=0.6140
Best Parameters: {'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 100, 'subsample': 1.0}

In [ ]:
# Load training and test datasets
train_df = pd.read_csv('a1_dataset/train_v9rqX0R.csv')
test_df = pd.read_csv('a1_dataset/test_AbJTz2l.csv')

# Preprocessing function
def preprocess(df):
    # Handle missing values
    df['Item_Weight'] = df['Item_Weight'].fillna(df['Item_Weight'].median())
    df['Outlet_Size'] = df['Outlet_Size'].fillna(df['Outlet_Size'].mode()[0])

    # Normalize Item_Fat_Content
    df['Item_Fat_Content'] = df['Item_Fat_Content'].replace({'low fat': 'Low Fat', 'LF': 'Low Fat', 'reg': 'Regular'})

    # Encode categorical variables
    cat_cols = df.select_dtypes(include=['object']).columns
    le = LabelEncoder()
    for col in cat_cols:
        df[col] = le.fit_transform(df[col])
    return df

# Preprocess train and test data
train_df = preprocess(train_df)
test_df_copy = test_df.copy()  # Keep original for output

test_df = preprocess(test_df)

# Features and target for training
X_train = train_df.drop(['Item_Outlet_Sales', 'Item_Identifier', 'Outlet_Identifier'], axis=1)
y_train = train_df['Item_Outlet_Sales']

# Prepare test features (drop identifiers)
X_test = test_df.drop(['Item_Identifier', 'Outlet_Identifier'], axis=1)

# Train XGBoost with best parameters
best_params = {'learning_rate': 0.05, 'max_depth': 3, 'n_estimators': 100, 'subsample': 1.0}
model = xgb.XGBRegressor(**best_params, random_state=42, verbosity=0)
model.fit(X_train, y_train)

# Predict on test data
test_predictions = model.predict(X_test)

# Append predictions to original test dataframe
test_df_copy['Item_Outlet_Sales_Predicted'] = test_predictions

# Save to CSV
test_df_copy.to_csv('test_predictions.csv', index=False)

print("Predictions saved to test_predictions.csv")
print(test_df_copy.head())

Predictions saved to test_predictions.csv


,Item_Identifier,Item_Weight,Item_Fat_Content,Item_Visibility,Item_Type,Item_MRP,Outlet_Identifier,Outlet_Establishment_Year,Outlet_Size,Outlet_Location_Type,Outlet_Type,Item_Outlet_Sales_Predicted
0,FDW58,20.750,Low Fat,0.007565,Snack Foods,107.8622,OUT049,1999,Medium,Tier 1,Supermarket Type1,1671.587036
1,FDW14,8.300,reg,0.038428,Dairy,87.3198,OUT017,2007,NaN,Tier 2,Supermarket Type1,1402.553833
2,NCN55,14.600,Low Fat,0.099575,Others,241.7538,OUT010,1998,NaN,Tier 3,Grocery Store,634.664551
3,FDQ58,7.315,Low Fat,0.015388,Snack Foods,155.0340,OUT017,2007,NaN,Tier 2,Supermarket Type1,2490.142578
4,FDY38,NaN,Regular,0.118599,Dairy,234.2300,OUT027,1985,Medium,Tier 3,Supermarket Type3,6061.709961
